arangod/Replication2/ReplicatedLog/ParticipantsCacheFeature.cpp

```diff
@@ -0,0 +1,114 @@
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Alex Petenchea
+/// @author Manuel Pöter
+////////////////////////////////////////////////////////////////////////////////
+
+#include "ParticipantsCacheFeature.h"
+
+#include "ApplicationFeatures/ApplicationServer.h"
+#include "ApplicationFeatures/CommunicationFeaturePhase.h"
+#include "FeaturePhases/DatabaseFeaturePhase.h"
+#include "Cluster/AgencyCallbackRegistry.h"
+#include "Cluster/AgencyCallback.h"
+#include "Cluster/ServerState.h"
+#include "Logger/LogMacros.h"
+#include "FeaturePhases/ClusterFeaturePhase.h"
+#include "FeaturePhases/FinalFeaturePhase.h"
+#include "FeaturePhases/AgencyFeaturePhase.h"
+#include "FeaturePhases/ServerFeaturePhase.h"
+#include "Cluster/ClusterFeature.h"
+
+namespace arangodb::replication2 {
+
+const std::string_view ParticipantsCacheFeature::kParticipantsHealthPath =
+    "arango/Supervision/Health";
+
+ParticipantsCacheFeature::ParticipantsCacheFeature(
+    application_features::ApplicationServer& server)
+    : ApplicationFeature(server, "ParticipantsCache") {
+  agencyCallback = std::make_shared<AgencyCallback>(
+      server, std::string(kParticipantsHealthPath),
+      [self = weak_from_this()](VPackSlice const& result) {
+        auto watcher = self.lock();
+        if (watcher) {
+          if (result.isNone()) {
+            LOG_DEVEL << "result is None";
+          } else {
+            LOG_DEVEL << result.toString();
+          }
+        } else {
+          LOG_DEVEL << "watcher destroyed";
+        }
+        return true;
+      },
+      true, false);
+
+  setOptional(true);
+  startsAfter<application_features::CommunicationFeaturePhase>();
+  startsAfter<application_features::DatabaseFeaturePhase>();
+  startsAfter<application_features::ClusterFeaturePhase>();
+  startsAfter<application_features::FinalFeaturePhase>();
+  startsAfter<application_features::AgencyFeaturePhase>();
+  startsAfter<application_features::ServerFeaturePhase>();
+  startsAfter<ClusterFeature>();
+}
+
+ParticipantsCacheFeature::~ParticipantsCacheFeature() {
+  try {
+    auto x = server().getFeature<ClusterFeature>().agencyCallbackRegistry();
+    x->unregisterCallback(agencyCallback);
+  } catch (std::exception const& ex) {
+    LOG_TOPIC("42af2", WARN, Logger::REPLICATION2)
+        << "caught unexpected exception while unregistering agency callback in "
+           "ParticipantsCacheFeature: "
+        << ex.what();
+  }
+}
```

Comment:
I suggest to deregister the call back in the features `stop` function. You have no control over when the destructor is executed with respect to other features destructor or the server destructor(afaik)